### PR TITLE
Orca: update the datatype of sparse features in NCF examples from Int to Long

### PR DESCRIPTION
**Pull request by zpeng1898**
1. Why the change?
To be compatible with `nn.embedding` in both [torch==1.7.1](https://pytorch.org/docs/1.7.1/generated/torch.nn.Embedding.html#embedding) and [torch==1.13.1](https://pytorch.org/docs/1.13/generated/torch.nn.Embedding.html?highlight=nn+embedding#torch.nn.Embedding)

image

image

### CODE REVIEW

**Review by hkvision — APPROVED** [review]

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 1]
```diff
@@ -17,31 +17,34 @@
 
 from pyspark.ml.feature import StringIndexer, MinMaxScaler, VectorAssembler
 from pyspark.ml import Pipeline
-from pyspark.sql.types import StructField, StructType, IntegerType, ArrayType, StringType
+from pyspark.sql.types import StructField, StructType, IntegerType, LongType, ArrayType, StringType
 from pyspark.sql.functions import udf, lit, collect_list, explode
 
 from bigdl.orca import OrcaContext
 
+# user and item ids are converted to LongType to be compatible with lower versions of PyTorch
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:25]: add `user and item ids and categorical features`?

File: python/orca/tutorial/NCF/process_xshards.py [hunk 2]
```diff
@@ -35,7 +38,7 @@ def ng_sampling(data, user_num, item_num, num_ng):
     data_X = data.values.tolist()
 
     # calculate a dok matrix
-    train_mat = sp.dok_matrix((user_num, item_num), dtype=np.int32)
+    train_mat = sp.dok_matrix((user_num, item_num), dtype=np.int64)
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_xshards.py:41]: is this change necessary?

File: python/orca/tutorial/NCF/process_xshards.py [hunk 3]
```diff
@@ -89,12 +93,18 @@ def prepare_data(dataset_dir, num_ng=4):
 
     print("Processing features...")
     # Categorical encoding
+    def convert_datatype(shard, col):
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_xshards.py:96]: rename to `convert_to_long`?

File: python/orca/tutorial/NCF/pytorch_dataset.py [hunk 4]
```diff
@@ -136,10 +138,10 @@ def process_ratings(dataset_dir, user_num, item_num):
     ratings = pd.read_csv(
         os.path.join(dataset_dir, "ratings.dat"),
         sep="::", header=None, names=["user", "item"],
-        usecols=[0, 1], dtype={0: np.int32, 1: np.int32})
+        usecols=[0, 1], dtype={0: np.int64, 1: np.int64})
 
     # load ratings as a dok matrix
-    train_mat = sp.dok_matrix((user_num, item_num), dtype=np.int32)
+    train_mat = sp.dok_matrix((user_num, item_num), dtype=np.int64)
```
Comment (hkvision) [inline python/orca/tutorial/NCF/pytorch_dataset.py:144]: is this necessary?

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 5]
```diff
@@ -17,31 +17,34 @@
 
 from pyspark.ml.feature import StringIndexer, MinMaxScaler, VectorAssembler
 from pyspark.ml import Pipeline
-from pyspark.sql.types import StructField, StructType, IntegerType, ArrayType, StringType
+from pyspark.sql.types import StructField, StructType, IntegerType, LongType, ArrayType, StringType
 from pyspark.sql.functions import udf, lit, collect_list, explode
 
 from bigdl.orca import OrcaContext
 
+# user and item ids and sparse features are converted to LongType to be compatible with
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:25]: user/item ids?

File: python/orca/tutorial/NCF/process_xshards.py [hunk 6]
```diff
@@ -88,13 +92,20 @@ def prepare_data(dataset_dir, num_ng=4):
     item_num = max(item_set) + 1
 
     print("Processing features...")
+
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_xshards.py:95]: remove this line?

File: python/orca/tutorial/NCF/process_xshards.py [hunk 7]
```diff
@@ -88,13 +92,20 @@ def prepare_data(dataset_dir, num_ng=4):
...
     # Categorical encoding
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_xshards.py:96]: put this comment together with the for loop

File: python/orca/tutorial/NCF/pytorch_dataset.py [hunk 8]
```diff
@@ -83,27 +86,28 @@ def __getitem__(self, idx):
...
+    sparse_features = ["gender", "zipcode", "category", "occupation"]
     dense_features = ["age"]
 
     users = pd.read_csv(
         os.path.join(dataset_dir, "users.dat"),
         sep="::", header=None, names=["user", "gender", "age", "occupation", "zipcode"],
         usecols=[0, 1, 2, 3, 4],
-        dtype={0: np.int32, 1: str, 2: np.int32, 3: np.int32, 4: str})
+        dtype={0: np.int64, 1: str, 2: np.int32, 3: np.int64, 4: str},
+        engine="python")
```
Comment (hkvision) [inline python/orca/tutorial/NCF/pytorch_dataset.py:97]: why adding this?

Comment (zpeng1898) [reply python/orca/tutorial/NCF/pytorch_dataset.py:97]: According to [the pandas document](https://pandas.pydata.org/pandas-docs/version/0.24/reference/api/pandas.errors.ParserWarning.html), When `sep` is more than a single character (e.g. regex separators), several warning lines will show:  python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'. 
So `engine="python"` will hide warnings.

File: python/orca/tutorial/NCF/process_xshards.py [hunk 9]
```diff
@@ -88,13 +92,21 @@ def prepare_data(dataset_dir, num_ng=4):
     item_num = max(item_set) + 1
 
     print("Processing features...")
+
+
+    def convert_to_long(shard, col):
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_xshards.py:97]: rename shard to df in the argument name?

File: python/orca/tutorial/NCF/process_xshards.py [hunk 10]
```diff
@@ -88,13 +92,21 @@ def prepare_data(dataset_dir, num_ng=4):
...
+        shard[col] = shard[col].astype(np.int64)
+        return shard
+
     # Categorical encoding
     for col in sparse_features[:-1]:  # occupation is already indexed.
         indexer = StringIndexer(inputCol=col)
         if col in users.get_schema()["columns"]:
             users = indexer.fit_transform(users)
+            users = users.transform_shard(lambda shard: convert_to_long(shard, col))
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_xshards.py:106]: rename shard to df?
Also for other transform_shard in this file.
